pyccel/ast/variable.py

```diff
@@ -0,0 +1,848 @@
+        (which can be quite long). This function therefore
+        replaces those expressions with calls to
+        PyccelArraySize
+        """
+        if not hasattr(shape,'__iter__'):
+            shape = [shape]
+
+        new_shape = []
+        for i,s in enumerate(shape):
+            if isinstance(s,(LiteralInteger, PyccelArraySize)):
+                new_shape.append(s)
+            elif isinstance(s, int):
+                new_shape.append(LiteralInteger(s))
+            elif s is None or isinstance(s,(Variable, Slice, PyccelAstNode, Function)):
+                new_shape.append(PyccelArraySize(self, i))
+            else:
+                raise TypeError('shape elements cannot be '+str(type(s))+'. They must be one of the following types: Integer(pyccel),'
+                                'Variable, Slice, PyccelAstNode, Integer(sympy), int, Function')
+        return tuple(new_shape)
+
+    @property
+    def name(self):
+        """ Name of the variable
+        """
+        return self._name
+
+    @property
+    def alloc_shape(self):
+        """ Shape of the variable at allocation
+
+        The shape used in pyccel is usually simplified to contain
+        only Literals and PyccelArraySizes but the shape for
+        the allocation of x cannot be `Shape(x)`
+        """
+        return self._alloc_shape
+
+    @property
+    def allocatable(self):
+        """ Indicates whether a Variable has a dynamic size
+        """
+        return self._allocatable
+
+    @allocatable.setter
+    def allocatable(self, allocatable):
+        if not isinstance(allocatable, bool):
+            raise TypeError('allocatable must be a boolean.')
+        self._allocatable = allocatable
+
+    @property
+    def cls_base(self):
+        """ Class from which the Variable inherits
+        """
+        return self._cls_base
+
+    @property
+    def is_const(self):
+        """ Indicates if the Variable is constant
+        within its context
+        """
+        return self._is_const
+
+    @property
+    def is_pointer(self):
+        """ Indicates if the Variable is a label for
+        something which points to another object.
+        In other words, the Variable does not own its data
+        """
+        return self._is_pointer
+
+    @is_pointer.setter
+    def is_pointer(self, is_pointer):
+        if not isinstance(is_pointer, bool):
+            raise TypeError('is_pointer must be a boolean.')
+        self._is_pointer = is_pointer
+
+    @property
+    def is_target(self):
+        """ Indicates if the data in this Variable is
+        shared with (pointed at by) another Variable
+        """
+        return self._is_target
+
+    @is_target.setter
+    def is_target(self, is_target):
+        if not isinstance(is_target, bool):
+            raise TypeError('is_target must be a boolean.')
+        self._is_target = is_target
+
+    @property
+    def is_polymorphic(self):
+        return self._is_polymorphic
```

Comment:
@saidctb this needs a docstring. Can you provide more information? I am not sure what a polymorphic variable is. This seems like something that should be the property of the dtype?

Reply:
it's used for classes,  but I'm not sure if it's used at all

Reply:
It is used once in the fortran printer but it seems to also exist in the dtype. I think I will leave it just in the dtype if that is ok with you

Reply:
yes fine